src/loader/bgw_launcher.c

```diff
@@ -481,21 +506,24 @@ message_restart_action(HTAB *db_htab, BgwMessage *message, VirtualTransactionId
 		terminate_background_worker(db_he->db_scheduler_handle);
 		wait_for_background_worker_shutdown(db_he->db_scheduler_handle);
 	}
-	else if (!bgw_total_workers_increment())	/* we still need to increment
-												 * if we haven't found an
-												 * entry */
+	else
 	{
-		report_bgw_limit_exceeded();
-		bgw_message_send_ack(message, ACK_FAILURE);
-		return;
+		db_he->db_scheduler_handle = NULL;
+		db_he->incremented = false;
```

Comment:
Btw, it seems weird that the restart case can create new hash table entries (with `HASH_ENTER`). 

This function should probably use `HASH_FIND` since it should only be possible to restart something that already exists.

This is significant because if we know that this function will never create a new hash entry, we also never need to increment/decrement the counter. This assumes that we decide that the number of hash entries should always reflect the current number of increments the launcher has done.